// @luohc92/vue3-image-viewer/dist/vue3-image-viewer.es.ts
import { defineComponent as J, reactive as K, ref as Q, computed as v, onMounted as Z, nextTick as ee, toRefs as ne, openBlock as w, createBlock as oe, Transition as A, withCtx as F, withDirectives as k, createElementVNode as a, normalizeStyle as b, createElementBlock as p, createCommentVNode as S, vShow as y, normalizeClass as L, Fragment as O, createVNode as R, renderList as te, toDisplayString as Y, render as _ } from "vue";
const $ = typeof window > "u", I = function() {
  return $ ? function(e, n, i) {
    e && n && i && e.attachEvent("on" + n, i);
  } : function(e, n, i) {
    e && n && i && e.addEventListener(n, i, !1);
  };
}(), z = function() {
  return $ ? function(e, n, i) {
    e && n && e.detachEvent("on" + n, i);
  } : function(e, n, i) {
    e && n && e.removeEventListener(n, i, !1);
  };
}(), se = function() {
  return !$ && !!window.navigator.userAgent.match(/firefox/i);
};
function B(e) {
  let n = !1;
  return function(...i) {
    n || (n = !0, window.requestAnimationFrame((c) => {
      e.apply(this, i), n = !1;
    }));
  };
}
const ae = J({
  name: "ImageViewer",
  props: {
    curIndex: {
      type: Number,
      default: 0
    },
    images: {
      type: Array,
      default: []
    },
    showDownload: {
      type: Boolean,
      default: !1
    },
    showThumbnail: {
      type: Boolean,
      default: !1
    },
    handlePosition: {
      type: String,
      default: "bottom"
    },
    onClose: {
      type: Function,
      default: () => {
      }
    },
    onDownload: {
      type: Function,
      default: (e) => {
      }
    },
    zIndex: {
      type: Number,
      default: 2e3
    },
    maskBgColor: {
      type: String,
      default: "rgba(0,0,0,0.5)"
    }
  },
  emits: ["destroy"],
  setup(e, { emit: n }) {
    const i = {
      CONTAIN: {
        name: "contain",
        icon: "icon-full-screen"
      },
      ORIGINAL: {
        name: "original",
        icon: "icon-c-scale-to-original"
      }
    }, c = se() ? "DOMMouseScroll" : "mousewheel", o = K({
      mode: i.CONTAIN,
      visible: !1,
      curIndex: 0,
      isMouseEnterThumbnail: !1,
      thumbnailTransitionShow: !1,
      transform: {
        scale: 1,
        deg: 0,
        rotateY: 0,
        offsetX: 0,
        offsetY: 0,
        enableTransition: !1
      },
      dragHandler: () => {
      },
      keyDownHandler: (t) => {
      },
      mouseWheelHandler: () => {
      }
    }), g = Q(), s = v(() => e.handlePosition == "bottom" ? {
      top: "10px",
      bottom: "auto"
    } : {
      bottom: "10px",
      top: "auto"
    }), h = v(() => e.handlePosition == "bottom" ? {
      bottom: e.showThumbnail && e.images && e.images.length > 1 ? "100px" : "30px",
      top: "auto"
    } : {
      top: "30px",
      bottom: "auto"
    }), E = v(() => {
      const { scale: t, deg: l, rotateY: u, offsetX: d, offsetY: r, enableTransition: m } = o.transform, D = {
        transform: `scale(${t}) rotate(${l}deg) rotateY(${u}deg)`,
        transition: m ? "transform .3s" : "",
        "margin-left": `${d}px`,
        "margin-top": `${r}px`,
        maxWidth: "",
        maxHeight: ""
      };
      return o.mode.name === i.CONTAIN.name && (D.maxWidth = D.maxHeight = "100%"), D;
    }), V = v(() => o.curIndex === 0), W = v(() => o.curIndex === e.images.length - 1), X = (t) => {
      o.isMouseEnterThumbnail = t;
    }, C = (t) => {
      if (!e.showThumbnail)
        return;
      o.curIndex = t;
      let l = g.value;
      if (!l)
        return;
      let u = l.clientWidth, d = l.scrollLeft, r = l.children[t].offsetLeft;
      u - r < 160 ? u - r < 0 && d === 0 ? l.scrollLeft = r - u + 160 : l.scrollLeft = d + 80 : r - d < 80 && (l.scrollLeft = r - 120);
    }, q = (t) => {
      if (t.button !== 0)
        return;
      const { offsetX: l, offsetY: u } = o.transform, d = t.pageX, r = t.pageY;
      o.dragHandler = B((m) => {
        o.transform.offsetX = l + m.pageX - d, o.transform.offsetY = u + m.pageY - r;
      }), I(document, "mousemove", o.dragHandler), I(document, "mouseup", () => {
        z(document, "mousemove", o.dragHandler);
      }), t.preventDefault();
    }, H = () => {
      if (V.value)
        return;
      const t = e.images.length;
      o.curIndex = (o.curIndex - 1 + t) % t, C(o.curIndex);
    }, M = () => {
      if (W.value)
        return;
      const t = e.images.length;
      o.curIndex = (o.curIndex + 1) % t, C(o.curIndex);
    }, P = () => {
      o.keyDownHandler = (t) => {
        switch (t.preventDefault(), t.keyCode) {
          case 17:
            f("mirror");
            break;
          case 18:
            f("clockwise");
            break;
          case 27:
            T();
            break;
          case 32:
            N();
            break;
          case 37:
            H();
            break;
          case 38:
            f("zoomIn");
            break;
          case 39:
            M();
            break;
          case 40:
            f("zoomOut");
            break;
        }
      }, o.mouseWheelHandler = B((t) => {
        const l = t.wheelDelta ? t.wheelDelta : -t.detail;
        if (o.isMouseEnterThumbnail) {
          let u = -1, d = 1, r = 0;
          r = l > 0 ? u * 50 : d * 50, g.value && (g.value.scrollLeft = g.value.scrollLeft + r);
        } else
          l > 0 ? f("zoomIn", {
            zoomRate: 0.015,
            enableTransition: !1
          }) : f("zoomOut", {
            zoomRate: 0.015,
            enableTransition: !1
          });
      }), I(document, "keydown", o.keyDownHandler), I(document, c, o.mouseWheelHandler);
    }, j = () => {
      z(document, "keydown", o.keyDownHandler), z(document, c, o.mouseWheelHandler), o.keyDownHandler = (t) => {
      }, o.mouseWheelHandler = () => {
      };
    }, f = (t, l = {}) => {
      const { zoomRate: u, rotateDeg: d, enableTransition: r } = {
        zoomRate: 0.2,
        rotateDeg: 90,
        enableTransition: !0,
        ...l
      }, { transform: m } = o;
      switch (t) {
        case "zoomOut":
          m.scale > 0.2 && (m.scale = parseFloat(
            (m.scale - u).toFixed(3)
          ));
          break;
        case "zoomIn":
          m.scale = parseFloat((m.scale + u).toFixed(3));
          break;
        case "clockwise":
          m.deg += d;
          break;
        case "anticlockwise":
          m.deg -= d;
          break;
        case "mirror":
          m.rotateY += 180;
          break;
      }
      m.enableTransition = r;
    }, x = () => {
      o.transform = {
        scale: 1,
        deg: 0,
        rotateY: 0,
        offsetX: 0,
        offsetY: 0,
        enableTransition: !1
      };
    }, N = () => {
      const t = Object.keys(i), l = Object.values(i), d = (l.findIndex((r) => r.name == o.mode.name) + 1) % t.length;
      o.mode = l[d], x();
    }, G = () => {
      e.onDownload(e.images[o.curIndex]);
    }, U = () => {
      T();
    }, T = () => {
      n("destroy"), e.onClose(), j(), o.visible = !1, o.thumbnailTransitionShow = !1;
    };
    return Z(() => {
      o.curIndex = e.curIndex, P(), o.visible = !0, ee(() => {
        o.thumbnailTransitionShow = !0;
      });
    }), {
      ...ne(o),
      close: T,
      handleTapClose: U,
      pre: H,
      next: M,
      thumbnailRef: g,
      imgStyle: E,
      handleMouseDown: q,
      sequenceStyle: s,
      actionStyle: h,
      handleActions: f,
      download: G,
      toggleMode: N,
      change: C,
      mouseEnterThumbnail: X
    };
  }
});
const ie = (e, n) => {
  const i = e.__vccOpts || e;
  for (const [c, o] of n)
    i[c] = o;
  return i;
}, le = /* @__PURE__ */ a("i", { class: "iconfont icon-close" }, null, -1), re = [
  le
], de = /* @__PURE__ */ a("i", { class: "iconfont icon-arrow-left" }, null, -1), me = [
  de
], ue = /* @__PURE__ */ a("i", { class: "iconfont icon-arrow-right" }, null, -1), ce = [
  ue
], fe = { class: "tmd-image-viewer__image" }, ge = ["src"], we = { class: "tmd-image-viewer_actions__inner" }, he = /* @__PURE__ */ a("span", { class: "tmd-image-viewer__actions__divider" }, null, -1), ve = /* @__PURE__ */ a("span", { class: "tmd-image-viewer__actions__divider" }, null, -1), be = /* @__PURE__ */ a("span", { class: "tmd-image-viewer__actions__divider" }, null, -1), pe = ["onClick"], ke = ["src"];
function ye(e, n, i, c, o, g) {
  return w(), oe(A, { name: "viewer-fade" }, {
    default: F(() => [
      k(a("div", {
        tabindex: "-1",
        class: "tmd-image-viewer__wrapper",
        style: b(`z-index:${e.zIndex}`)
      }, [
        a("div", {
          class: "tmd-image-viewer__mask",
          style: b(`background-color:${e.maskBgColor};`),
          onClick: n[0] || (n[0] = (...s) => e.handleTapClose && e.handleTapClose(...s))
        }, null, 4),
        a("span", {
          class: "tmd-image-viewer__btn tmd-image-viewer__close",
          onClick: n[1] || (n[1] = (...s) => e.close && e.close(...s))
        }, re),
        e.images && e.images.length > 1 ? (w(), p("span", {
          key: 0,
          class: "tmd-image-viewer__btn tmd-image-viewer__pre",
          onClick: n[2] || (n[2] = (...s) => e.pre && e.pre(...s))
        }, me)) : S("", !0),
        e.images && e.images.length > 1 ? (w(), p("span", {
          key: 1,
          class: "tmd-image-viewer__btn tmd-image-viewer__next",
          onClick: n[3] || (n[3] = (...s) => e.next && e.next(...s))
        }, ce)) : S("", !0),
        a("div", fe, [
          k(a("img", {
            src: e.images[e.curIndex],
            style: b(e.imgStyle),
            onMousedown: n[4] || (n[4] = (...s) => e.handleMouseDown && e.handleMouseDown(...s))
          }, null, 44, ge), [
            [y, e.images && e.images.length > 0 && e.images[e.curIndex]]
          ])
        ]),
        a("div", {
          class: "tmd-image-viewer__actions",
          style: b(e.actionStyle)
        }, [
          a("div", we, [
            a("span", {
              class: "iconfont icon-zoom-out",
              onClick: n[5] || (n[5] = (s) => e.handleActions("zoomOut"))
            }),
            a("span", {
              class: "iconfont icon-zoom-in",
              onClick: n[6] || (n[6] = (s) => e.handleActions("zoomIn"))
            }),
            he,
            a("span", {
              class: "iconfont icon-refresh-left",
              onClick: n[7] || (n[7] = (s) => e.handleActions("anticlockwise"))
            }),
            a("span", {
              class: "iconfont icon-refresh-right",
              onClick: n[8] || (n[8] = (s) => e.handleActions("clockwise"))
            }),
            a("span", {
              class: "iconfont icon-mirror",
              onClick: n[9] || (n[9] = (s) => e.handleActions("mirror"))
            }),
            ve,
            a("span", {
              class: L(["iconfont", e.mode.icon]),
              onClick: n[10] || (n[10] = (...s) => e.toggleMode && e.toggleMode(...s))
            }, null, 2),
            e.showDownload ? (w(), p(O, { key: 0 }, [
              be,
              a("span", {
                class: "iconfont icon-download",
                onClick: n[11] || (n[11] = (...s) => e.download && e.download(...s))
              })
            ], 64)) : S("", !0)
          ])
        ], 4),
        R(A, {
          name: "thumbnail-fade",
          onAfterEnter: n[14] || (n[14] = (s) => e.change(e.curIndex))
        }, {
          default: F(() => [
            k(a("div", {
              class: "tmd-image-viewer__thumbnail",
              ref: "thumbnailRef",
              onMouseenter: n[12] || (n[12] = (s) => e.mouseEnterThumbnail(!0)),
              onMouseleave: n[13] || (n[13] = (s) => e.mouseEnterThumbnail(!1))
            }, [
              (w(!0), p(O, null, te(e.images, (s, h) => (w(), p("div", {
                class: L(["tmd-image-viewer__thumbnail__inner", { current: e.curIndex === h }]),
                key: s + "_thumbnail_" + h,
                onClick: (E) => e.change(h)
              }, [
                a("img", { src: s }, null, 8, ke)
              ], 10, pe))), 128))
            ], 544), [
              [
                y,
                e.thumbnailTransitionShow && e.showThumbnail && e.images && e.images.length > 1
              ]
            ])
          ]),
          _: 1
        }),
        k(a("span", {
          class: "tmd-image-viewer__sequence",
          style: b(e.sequenceStyle)
        }, Y(e.curIndex + 1) + " / " + Y(e.images.length), 5), [
          [y, !e.showThumbnail]
        ])
      ], 4), [
        [y, e.visible]
      ])
    ]),
    _: 1
  });
}
const Ie = /* @__PURE__ */ ie(ae, [["render", ye]]), Ce = {
  curIndex: 0,
  images: [],
  showDownload: !1,
  showThumbnail: !1,
  handlePosition: "bottom",
  onClose: Function,
  onDownload: Function,
  zIndex: 2e3,
  maskBgColor: "rgba(0,0,0,0.5)"
}, De = (e) => {
  const n = document.createElement("div");
  n.className = "tmd-image-viewer-container";
  const i = {
    ...Ce,
    ...e
  }, c = R(Ie, i);
  c && c.props && (c.props.onDestroy = () => {
    setTimeout(() => {
      _(null, n);
    }, 300);
  }), _(c, n), n.firstElementChild && document.body.appendChild(n.firstElementChild);
};
export {
  De as default
};
